File: LeetCode/110. Balanced Binary Tree/balanced_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        return self.helper(root)[0]
    
    def helper(self, node):
        if not node:
            return (True, 0)
        
        isleft, heightleft = self.helper(node.left)
        isright, heightright = self.helper(node.right)
        
        if isleft and isright and abs(heightleft - heightright) < 2:
            return (True, max(heightright, heightleft) + 1)
        
        return (False, 0)

File: LeetCode_cpp/110. Balanced Binary Tree/balanced_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int helper(TreeNode* node, bool *isbt) {
        if (node->left == null and node->right == null){
            isbt = true;
            return 0;
        }
        
        bool isr = false;
        bool isl = false;
        int rh = helper(node->right, isr);
        int rl = helper(node->left, isl);
        
        if (isr == false or isl == false) return 0;
        int dif = rh - rl;
        if (dif < 0) dif = dif * (-1);
        if (dif == 0 or dif == 1){
            isbt = true;
            return dif;
        }
        else{
            return 0;
        }
            
    }
    
    bool isBalanced(TreeNode* root) {
        bool isbt = false;
        int height = helper(root, isbt);
        return isbt;   
        
    }
};
